_posts/2022-01-01-the-kubernetes-discovery-cache-blessing-and-curse.md

```diff
@@ -0,0 +1,101 @@
+
+So while the discovery cache is actually adding usability to Kubernetes, it also is the limiting factor for extending the platform with custom controllers and CRDs.
+
+### Scaling
+
+Especially the crossplane community has a vested interest in unlocking this limitation because crossplane's entire design philosophy is built upon the idea to create CRDs for every object in the real world and reconciling it through controllers. But it will also be important for other controllers introducing many CRDs like the [GCP Config Connector](https://github.com/GoogleCloudPlatform/k8s-config-connector) or the [Azure Service Operator](https://github.com/Azure/azure-service-operator).
+
+For now the [aforementioned issue on kubernetes/kubernetes](https://github.com/kubernetes/kubernetes/issues/105489) based on my user report regarding many `GET` requests after a simple `kubectl get pods` triggered a set of PRs ([1](https://github.com/kubernetes/kubernetes/pull/105520), [2](https://github.com/kubernetes/kubernetes/pull/107131)) aimed at increasing the rate limits during discovery. However, this is just kicking the can down the road (or as [@liggitt correctly put it](https://github.com/kubernetes/kubernetes/pull/105520#discussion_r723535829) the 'kubernetes equivalent of the debt ceiling') as it's not solving the underlying issue of many unnecessary `GET` requests, but merely not rate limiting as often, which still means a strain on resources and that we will run into the same issue again at a later point in time with even more CRDs. While `kubectl` still performs 100s of `GET` requests, at least the total run time is [roughly cut in half](https://github.com/kubernetes/kubernetes/pull/107131#issue-1084329062) as there is no additional rate limiting anymore with the fixes.
+
+I also raised [a separate issue](https://github.com/kubernetes/kubernetes/issues/107130) to challenge the status quo of invalidating the cache every 10 minutes by increasing that default and also to make this timeout configurable (rather than [hard coding it](https://github.com/kubernetes/kubernetes/blob/0fb71846df9babb6012a7fce22e2533e9d795baa/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go#L253)). But again, this just raises limits and doesn't actually minimize the amount of unused `GET` requests.
+
+So the real, lasting solution might be a bit more involved and require to only `GET` the `serverresources.json` of a group version that is actually requested once the cache got invalid or isn't present. So a request for `kubectl get pods` would only populate the `~/.kube/cache/discovery/<host>/v1/serverresources.json` file (because `pods` are in group `""` and version `v1`) rather than every single group version. This would eliminate all unnecessary requests for unrelated resources and majorly reduce the total amount of `GET` requests. This solution would also require a server side change to offer and endpoint that reveals all potential group versions for a given kind.
```

Comment:
```suggestion
So the real, lasting solution might be a bit more involved and require to only `GET` the `serverresources.json` of a group version that is actually requested once the cache gets invalid or isn't present. So a request for `kubectl get pods` would only populate the `~/.kube/cache/discovery/<host>/v1/serverresources.json` file (because `pods` are in group `""` and version `v1`) rather than every single group version. This would eliminate all unnecessary requests for unrelated resources and majorly reduce the total amount of `GET` requests. This solution would also require a server-side change to offer an endpoint that reveals all potential group versions for a given kind.
```